### PR TITLE
Add io.qt.QtCreator

### PR DESCRIPTION
**Pull request by 4c0n**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

**Comment by 4c0n** [comment]
bot, build io.qt.QtCreator

**Comment by 4c0n** [comment]
bot, build io.qt.QtCreator

File: io.qt.QtCreator.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    "rename-icon": "QtProject-qtcreator",
+    "finish-args": [
+        "--share=ipc",
+        "--socket=wayland",
+        "--socket=fallback-x11",
+        "--filesystem=host"
+    ],
+    "modules": [
+        {
+            "name": "clang",
```
Comment (barthalion) [inline io.qt.QtCreator.json:18]: Anything wrong with clang provided in the Sdk? It's not the latest but not that old either, `10.0.1`.

**Comment by barthalion** [comment]
```
builddir/files/share/appdata/io.qt.QtCreator.appdata.xml: FAILED:
• tag-missing           : <content_rating> required [use https://odrs.gnome.org/oars]
• style-invalid         : Not enough <screenshot> tags, minimum is 1
• tag-missing           : <release> required
```

**Comment by proletarius101** [comment]
Hi @4c0n are you still working on this? If no, I would like to work on this too.

**Comment by 4c0n** [comment]
@proletarius101 many things came up so had to put this on hold for now. You're very welcome to take over if you'd like 🙂 I don't remember exactly why I couldn't get it to build with the included clang unfortunately, but it's easy to find out just by removing the clang build stage and perhaps adjusting some arguments.

**Comment by 4c0n** [comment]
bot, build io.qt.QtCreator

**Comment by 4c0n** [comment]
bot, build io.qt.QtCreator

**Comment by 4c0n** [comment]
FYI: I'm trying to get the appdata changes merged in the upstream repository, however Qt has their own system, not sure how to submit the changes, so for now I opened this Jira issue: https://bugreports.qt.io/browse/QTCREATORBUG-25635

**Comment by 4c0n** [comment]
bot, build io.qt.QtCreator

File: io.qt.QtCreator.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{
+    "app-id": "io.qt.QtCreator",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.15",
+    "sdk": "org.kde.Sdk",
+    "command": "qtcreator",
+    "rename-desktop-file": "org.qt-project.qtcreator.desktop",
```
Comment (barthalion) [inline io.qt.QtCreator.json:7]: I'd say that if that's the ID used universally across qtcreator codebase, we should stick to it.

**Comment by 4c0n** [comment]
bot, build io.qt.QtCreator

**Comment by 4c0n** [comment]
bot, build io.qt.QtCreator

**Comment by 4c0n** [comment]
bot, build org.qt-project.QtCreator

**Comment by 4c0n** [comment]
bot, build org.qt-project.QtCreator

File: io.qt.QtCreator.json (hunk 2 above)
Comment (4c0n) [reply io.qt.QtCreator.json]: @barthalion I didn't remember why I had to change it, but now the error popped up  again:
`error: 'org.qt-project.QtCreator' is not a valid application name: Only last name segment can contain -`

What would you like me to do?

**Comment by barthalion** [comment]
Ah, I guess if you need to rename it due to dash in the middle of ID section, you can just revert that change and we can go with modified ID.

**Comment by barthalion** [comment]
FYI I also modified your PR's description to include the checklist we ask submitters to fill.

**Comment by 4c0n** [comment]
Submitted the changes made by the patches to the upstream:
https://codereview.qt-project.org/c/qt-creator/qt-creator/+/346311

**Comment by 4c0n** [comment]
Contacted upstream devs:
https://bugreports.qt.io/browse/QTCREATORBUG-25635

**Comment by 4c0n** [comment]
bot, build io.qt.QtCreator

**Comment by 4c0n** [comment]
> FYI I also modified your PR's description to include the checklist we ask submitters to fill.

Thanks I did my best to fulfill it.

**Comment by barthalion** [comment]
Hi @marbub, is it something you or someone else from QT Creator team would be interested in co-maintaining?

**Comment by marbub** [comment]
Hello @barthalion, [I worked on a flatpak for Qt Creator some years ago](https://git.qt.io/mabubke/flatpaks). Some of the stuff is outdated but you have to apply same patches.

**Comment by marbub** [comment]
But to this time it was not possible to detect compiler in the host usr. Something many developer wanted.

**Comment by barthalion** [comment]
Yeah, this is still not easily doable. Application can claim `--talk-name=org.freedesktop.Flatpak` and that lets users configure custom commands inside IDE to run `flatpak-spawn --host g++` that will run `g++` on the host, not inside the sandbox.

**Comment by marbub** [comment]
Have you looked into the [Qt Creator Readme](https://code.qt.io/cgit/qt-creator/qt-creator.git/tree/README.md)? cmake is now the official build system for QtCreator. There is a patched LLVM too. I need some patches on top of clang-format.

**Comment by 4c0n** [comment]
@marbub Thank you for your response and comments. It would seem that you have some valuable insights in how Qt Creator works and that there are patches required for it to work properly as a flatpak. :mag: 

Are you or someone else from the contributors/maintainers interested in co-maintaining this manifest? That way you could potentially provide these patches and we could document why they are required for example. Perhaps setup a testing procedure?

Otherwise, of course you're welcome to suggest/introduce changes, however we need to start somewhere and keeping track of issues and changes/maintaining will become a lot easier once this is approved as the manifest will reside in its own repository.

Please see https://github.com/flathub/flathub/wiki/App-Submission for more information on how that works :smiley:

**Comment by marbub** [comment]
Hello @4c0n , do have read the [Readme](https://code.qt.io/cgit/qt-creator/qt-creator.git/tree/README.md)? It says how to build QtCreator and where the repository with the extra patches is. I am now working on Qt DesignStudio, which is QtCreator compiled with different compile flags which produces a different user interface. I could help with that.

**Comment by 4c0n** [comment]
@marbub Alright so you're not interested in co-maintaining this manifest? For Qt DesignStudio I think it would be more appropriate for it to be available as a separate flatpak.

I have read the readme. It actually states that LLVM/Clang is an optional dependency and also that the recommended build tool would be Ninja, although cmake is still supported.

**Comment by 4c0n** [comment]
@marbub thanks for clarifying that part. I don't mean to be rude, but you have kind of avoided answering the question 3 times now. Would you be so kind as to grant us with an answer?

**Comment by marbub** [comment]
It really depends what co maintaining means. I like to be reactively but not proactively involved.

**Comment by barthalion** [comment]
Yeah, I think that's fine. We can grant you write access in case you want to close issues/merge pull requests/push changes but there's no obligation you really have to.

**Comment by marbub** [comment]
Okay, so it is fine for me.

**Comment by marbub** [comment]
@tjenssen is working with QtCreator/DesignStudio build system. Could he help too?

**Comment by 4c0n** [comment]
@barthalion Do we need anything else? :rocket:

**Comment by 4c0n** [comment]
FYI: The submitted patches for the appdata file seem to be as good as resolved, but I guess we'd have to wait until it ends up in a release until they're no longer required here:

https://codereview.qt-project.org/c/qt-creator/qt-creator/+/346311

**Comment by 4c0n** [comment]
For anyone interested:

Building the patched clang fork is a real pain in the butt, it takes a lot of resources and patience. So far I've got it to build, but unfortunately it was missing a few things and didn't get detected in the cmake-ninja build of qt-creator. Linking the binaries it seems can only be done with 1 thread otherwise it quickly eats through all available memory and swap space. It also requires a huge amount of disk space besides that.

Unfortunately I don't have enough memory/disk space for this kind of caching and I only have like an hour a day if lucky to work on it, but feel free to throw me some hints or help if you want here: 
https://github.com/4c0n/flathub/tree/io.qt.QtCreator-new

I'll transfer that branch over to the new repository when it gets to that.

**Comment by marbub** [comment]
You don't need to build everything. Only libClang and libFormat is needed.

**Comment by 4c0n** [comment]
@marbub I'm using https://github.com/4c0n/flathub/blob/ef181d2e979f76c0f8c66f25f7c301ab6cd7c0eb/io.qt.QtCreator.json#L31 to limit what is build, just like in the readme instructions. Do you know any other flags/options to limit it even more?

**Comment by marbub** [comment]
@4c0n Like you can see in [LLVm build instruction](https://llvm.org/docs/CMake.html) there are many options which can be disabled. Actually do not work on the C++ support anymore. So I have no configuration. But you can easily reduce the LLVM build with it.

**Comment by 4c0n** [comment]
It seems that https://llvm.org/docs/BuildingADistribution.html#options-for-reducing-size contains a lot more useful info on that.

Perhaps it could be useful to expand the docs on that in the Qt repositories, as with the standard instructions you'll have to sacrifice about 77GB of disk space... Also when using cmake it all of a sudden is not an optional dependency anymore apparently. :man_shrugging:

**Comment by barthalion** [comment]
bot, build io.qt.QtCreator

File: io.qt.QtCreator.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
+{
+    "app-id": "io.qt.QtCreator",
+    "runtime": "org.kde.Platform",
```
Comment (barthalion) [inline io.qt.QtCreator.json:3]:
```suggestion
    "runtime": "org.kde.Sdk",
```

File: io.qt.QtCreator.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    "sdk": "org.kde.Sdk",
+    "command": "qtcreator",
+    "rename-desktop-file": "org.qt-project.qtcreator.desktop",
+    "rename-appdata-file": "org.qt-project.qtcreator.appdata.xml",
+    "rename-icon": "QtProject-qtcreator",
+    "finish-args": [
+        "--share=ipc",
+        "--socket=wayland",
+        "--socket=fallback-x11",
+        "--filesystem=host"
```
Comment (barthalion) [inline io.qt.QtCreator.json:14]:
```suggestion
        "--filesystem=host",
        "--device=dri"
```

**Comment by 4c0n** [comment]
Thanks @barthalion !

**Pull request closed without merging**